src/hardware/shared_adc.rs

```diff
@@ -0,0 +1,113 @@
+/// Shared Internal ADC Support
+///
+/// # Description
+/// This module provides an abstraction to share ownership of a single ADC peripheral with multiple
+/// ADC channels attached to it.
+///
+/// The design of this module mimics that of [`shared-bus`].
+///
+/// First, the shared ADC is created with the use of a macro, which places the ADC peripheral into
+/// a mutable, static (singleton) location. Then, individual channels are created by passing in the
+/// associated ADC input pin to the [SharedAdc::create_channel()] function to generate an
+/// [AdcChannel]. The [AdcChannel]'s ownership can then be moved to any required drivers.
+///
+/// ## Synchronization
+/// If the multiple priorities utilize the ADC that results in resource pre-emption, pre-emption is
+/// protected against through the use of an atomic bool. Attempting to utilize the ADC from a
+/// higher priority level while it is in use at a lower level will result in a [AdcError::InUse].
+use embedded_hal::adc::{Channel, OneShot};
+use stm32h7xx_hal as hal;
+
+#[derive(Debug, Copy, Clone)]
+pub enum AdcError {
+    /// Indicates that the ADC channel has already been allocated.
+    Allocated,
+
+    /// Indicates that the ADC is already in use
+    InUse,
+}
+
+/// A single channel on an ADC peripheral.
+pub struct AdcChannel<'a, Adc, PIN> {
+    pin: PIN,
+    slope: f32,
+    mutex: &'a spinning::Mutex<hal::adc::Adc<Adc, hal::adc::Enabled>>,
+}
+
+impl<'a, Adc, PIN> AdcChannel<'a, Adc, PIN>
+where
+    PIN: Channel<Adc, ID = u8>,
+    hal::adc::Adc<Adc, hal::adc::Enabled>: OneShot<Adc, u32, PIN>,
+    <hal::adc::Adc<Adc, hal::adc::Enabled> as OneShot<Adc, u32, PIN>>::Error:
+        core::fmt::Debug,
+{
+    /// Read the ADC channel.
+    ///
+    /// # Returns
+    /// The normalized ADC measurement as a ratio of full-scale.
+    pub fn read(&mut self) -> Result<f32, AdcError> {
+        let mut adc = self.mutex.try_lock().ok_or(AdcError::InUse)?;
+        Ok(adc.read(&mut self.pin).unwrap() as f32 / self.slope)
+    }
+}
+
+/// An ADC peripheral that can provide ownership of individual channels for sharing between
+/// drivers.
+pub struct SharedAdc<Adc> {
+    mutex: spinning::Mutex<hal::adc::Adc<Adc, hal::adc::Enabled>>,
+    allocated_channels: core::cell::RefCell<[bool; 20]>,
```

Comment:
Is that needed? The `hal` will presumably ensure that `adc::Channel` is unique (it's the pin), and since that's owned by `AdcChannel`.

Reply:
There's some edge cases for internal channels where this isn't necessarily the case (although they are few and limited). One example is the CPU temperature sensor. You just instantiate a `let temp = hal::adc::Temperature::new()` to get an object that impls the adc::Channel.

That being said, I'm not opposed to just removing this check and making it up to the user to not do dumb things. For the vast majority of pin-related channels, as you note, this won't be an issue because of ownership.

Reply:
Ah. But even then, if multiple `adc::Temperature::new()` can be created, they will happily coexist, right?

Reply:
That's an interesting point. Indeed, it seems like there's no reason that an AdcChannel shouldn't be copyable if we're avoiding potential pre-emption of the resource.

Reply:
I think as long as you exclusively consume channels, you'd be fine. If we provided a `free()` mechanism that returned the pins, this obviously wouldn't work. In any case, that's out-of-scope here, so I'll go ahead and remove the checks. Should clean up the code a bit.